### ISSUE TITLE
fatal error: concurrent map read and map write

### ISSUE THREAD

**Issue by rlenk** [hashicorp/nomad#2668]
Nomad clients fail with `fatal error: concurrent map read and map write`

### Nomad version
Nomad v0.5.6

### Operating system and Environment details
The environment contains five Consul servers, three Nomad servers, 15 Nomad clients. There are 100 service jobs running consisting of eight containers each, and 200 periodic batch jobs consisting of three containers that run once per hour (spread over a 20 minute period).

* CentOS Linux release 7.2.1511 (Core)
* Consul v0.8.1
* Docker 1.13.1

### Issue
Nomad client agents fail three to four times daily on various nodes with the following error:

```
fatal error: concurrent map read and map write
```

### Nomad Client logs
Excerpt from the logs containing the error and panic goroutine trace (complete traces attached):

```
2017/05/23 17:02:18.127968 [WARN] client: error fetching stats of task adpod-caswriter: stats collection hasn't started yet
2017/05/23 17:02:19.128197 [WARN] client: error fetching stats of task adpod-caswriter: stats collection hasn't started yet
2017/05/23 17:02:28.553256 [INFO] client: task "features-casreader" for alloc "5b5b51c8-c08b-c382-0f5f-ef8c6014769c" completed successfully
2017/05/23 17:02:28.553329 [INFO] client: Not restarting task: features-casreader for alloc: 5b5b51c8-c08b-c382-0f5f-ef8c6014769c
2017/05/23 17:02:28.907668 [INFO] client: task "adpod" for alloc "5b5b51c8-c08b-c382-0f5f-ef8c6014769c" completed successfully
2017/05/23 17:02:28.907715 [INFO] client: Not restarting task: adpod for alloc: 5b5b51c8-c08b-c382-0f5f-ef8c6014769c
2017/05/23 17:02:28.923687 [INFO] client: task "adpod-caswriter" for alloc "5b5b51c8-c08b-c382-0f5f-ef8c6014769c" completed successfully
2017/05/23 17:02:28.923728 [INFO] client: Not restarting task: adpod-caswriter for alloc: 5b5b51c8-c08b-c382-0f5f-ef8c6014769c
2017/05/23 17:02:28.925330 [INFO] client: marking allocation 5b5b51c8-c08b-c382-0f5f-ef8c6014769c for GC
fatal error: concurrent map read and map write

goroutine 160783 [running]:
runtime.throw(0x128bd80, 0x21)
	/opt/go/src/runtime/panic.go:596 +0x95 fp=0xc4245d9538 sp=0xc4245d9518
runtime.mapaccess2(0x10ba040, 0xc42176c720, 0xc4212c1560, 0x4ed99c, 0xc420e66f20)
	/opt/go/src/runtime/hashmap.go:377 +0x244 fp=0xc4245d9580 sp=0xc4245d9538
reflect.mapaccess(0x10ba040, 0xc42176c720, 0xc4212c1560, 0xc42176c720)
	/opt/go/src/runtime/hashmap.go:1128 +0x3f fp=0xc4245d95b8 sp=0xc4245d9580
reflect.Value.MapIndex(0x10ba040, 0xc42176c770, 0x195, 0x104d020, 0xc4212c1560, 0x98, 0x112a920, 0xc421764700, 0x16)
	/opt/go/src/reflect/value.go:1052 +0x128 fp=0xc4245d9640 sp=0xc4245d95b8
encoding/json.(*mapEncoder).encode(0xc42011cb40, 0xc420e66f20, 0x10ba040, 0xc42176c770, 0x195, 0x100)
	/opt/go/src/encoding/json/encode.go:690 +0x534 fp=0xc4245d97b0 sp=0xc4245d9640
encoding/json.(*mapEncoder).(encoding/json.encode)-fm(0xc420e66f20, 0x10ba040, 0xc42176c770, 0x195, 0x100)
	/opt/go/src/encoding/json/encode.go:706 +0x64 fp=0xc4245d97f0 sp=0xc4245d97b0
encoding/json.(*structEncoder).encode(0xc42083e810, 0xc420e66f20, 0x11749e0, 0xc42176c750, 0x199, 0x100)
	/opt/go/src/encoding/json/encode.go:645 +0x253 fp=0xc4245d9950 sp=0xc4245d97f0
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc420e66f20, 0x11749e0, 0xc42176c750, 0x199, 0xc421760100)
	/opt/go/src/encoding/json/encode.go:659 +0x64 fp=0xc4245d9990 sp=0xc4245d9950
encoding/json.(*ptrEncoder).encode(0xc42011cb48, 0xc420e66f20, 0x11a15a0, 0xc423ea7a58, 0x196, 0x100)
	/opt/go/src/encoding/json/encode.go:786 +0xe3 fp=0xc4245d99d0 sp=0xc4245d9990
encoding/json.(*ptrEncoder).(encoding/json.encode)-fm(0xc420e66f20, 0x11a15a0, 0xc423ea7a58, 0x196, 0x100)
	/opt/go/src/encoding/json/encode.go:791 +0x64 fp=0xc4245d9a10 sp=0xc4245d99d0
encoding/json.(*structEncoder).encode(0xc4201d7c20, 0xc420e66f20, 0x11992e0, 0xc423ea7a40, 0x199, 0x100)
	/opt/go/src/encoding/json/encode.go:645 +0x253 fp=0xc4245d9b70 sp=0xc4245d9a10
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc420e66f20, 0x11992e0, 0xc423ea7a40, 0x199, 0xc423ea0100)
	/opt/go/src/encoding/json/encode.go:659 +0x64 fp=0xc4245d9bb0 sp=0xc4245d9b70
encoding/json.(*ptrEncoder).encode(0xc42011cba8, 0xc420e66f20, 0x1002840, 0xc423ea7a40, 0x16, 0x1000100)
	/opt/go/src/encoding/json/encode.go:786 +0xe3 fp=0xc4245d9bf0 sp=0xc4245d9bb0
encoding/json.(*ptrEncoder).(encoding/json.encode)-fm(0xc420e66f20, 0x1002840, 0xc423ea7a40, 0x16, 0xc423ea0100)
	/opt/go/src/encoding/json/encode.go:791 +0x64 fp=0xc4245d9c30 sp=0xc4245d9bf0
encoding/json.(*encodeState).reflectValue(0xc420e66f20, 0x1002840, 0xc423ea7a40, 0x16, 0x100)
	/opt/go/src/encoding/json/encode.go:323 +0x82 fp=0xc4245d9c68 sp=0xc4245d9c30
encoding/json.(*encodeState).marshal(0xc420e66f20, 0x1002840, 0xc423ea7a40, 0x10b0100, 0x0, 0x0)
	/opt/go/src/encoding/json/encode.go:296 +0xb8 fp=0xc4245d9ca0 sp=0xc4245d9c68
encoding/json.Marshal(0x1002840, 0xc423ea7a40, 0xc42001d350, 0xc42001d300, 0x4b, 0x0, 0xc4217d9d60)
	/opt/go/src/encoding/json/encode.go:161 +0x6e fp=0xc4245d9ce8 sp=0xc4245d9ca0
github.com/hashicorp/nomad/client.persistState(0xc423ea7b30, 0x4b, 0x1002840, 0xc423ea7a40, 0xc4217d9ebc, 0xc4217d9eb8)
	/opt/gopath/src/github.com/hashicorp/nomad/client/util.go:81 +0x50 fp=0xc4245d9e28 sp=0xc4245d9ce8
github.com/hashicorp/nomad/client.(*AllocRunner).saveAllocRunnerState(0xc422f36e10, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:245 +0x261 fp=0xc4245d9ef0 sp=0xc4245d9e28
github.com/hashicorp/nomad/client.(*AllocRunner).syncStatus(0xc422f36e10, 0xc4217bb860, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:346 +0x50 fp=0xc4245d9f18 sp=0xc4245d9ef0
github.com/hashicorp/nomad/client.(*AllocRunner).dirtySyncState(0xc422f36e10)
	/opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:334 +0xc1 fp=0xc4245d9fd8 sp=0xc4245d9f18
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4245d9fe0 sp=0xc4245d9fd8
created by github.com/hashicorp/nomad/client.(*AllocRunner).Run
	/opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:459 +0x8b
```

[nomad_err.txt](https://github.com/hashicorp/nomad/files/1023437/nomad_err.txt)

**Comment by dadgar** [comment]
@rlenk Hey sorry you are experiencing that! Nomad 0.6.0 (the next release) has pretty drastically changed the code path for saving state. Do you have a staging environment were you would be able to run the workload with a build from master to see if this has been fixed?

**Comment by rlenk** [comment]
@dadgar We can easily run with a build from master to see if this is still an issue. Do you have a timeline for the 0.6.0 release?

**Comment by dadgar** [comment]
[nomad.zip](https://github.com/hashicorp/nomad/files/1026640/nomad.zip)

@rlenk Wanna give this a try? It is built from master. Timeline is roughly 1 month.

**Comment by rlenk** [comment]
@dadgar Thanks! We will give that build a try. The failures have been intermittent, so it will likely be a few days before we can tell if there's an improvement.

**Comment by dadgar** [comment]
@rlenk Perfect. Yeah we have time to debug before cutting 0.6.0 so hopefully we can put the nail in the coffin with you helping debug!

**Comment by rlenk** [comment]
@dadgar I'm happy to report that this issue seems to be fixed in the 0.6.0dev build that you provided. We've been running for 5+ days with the same workload and zero client failures.

**Comment by dadgar** [comment]
@rlenk Awesome! Would you mind pinging this issue if it does come back with that build! Otherwise lets close this 👍
